### PR TITLE
Fix for #2593

### PR DESCRIPTION
**Pull request by mrtracy**
Finally got to the bottom of this, with special thanks to Tobias, Tamir and Ben.

The test which exercises this most consistently is TestStoreRangeRebalance; however, it was only able to quickly reproduce this in the multicpu branch, where it was reproduced with reliability.

A six node cluster was able to run stably under load for well over an hour with this fix; the bug appears to be completely gone.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3074)

### CODE REVIEW

**Comment by tamird** [comment]
LGTM. Seems reasonable to me that testing for this is Hard™.

File: multiraft/multiraft.go [hunk 1]
```diff
@@ -989,8 +982,21 @@ func (s *state) propose(p *proposal) {
...
+		// If this is an initial proposal, go ahead and remove the command. If
+		// this is a re-proposal, there is a good chance that the original
+		// proposal will still commit even though the node has been removed from
+		// the group.
+		//
+		// Therefore, we will not remove the pending command until the group
+		// itself is removed, the node is re-added to the group
+		// (which will trigger another re-proposal that will succeed), or the
+		// proposal is committed.
+		if initialProposal {
```
Comment (tschottdorf) [inline multiraft/multiraft.go]: isn't `initialProposal` equal to the `ok` in `_, ok := g.pending[p.commandID]` and thus not required?

**Comment by tschottdorf** [comment]
LGTM mod my question.

**Comment by bdarnell** [comment]
LGTM mod @tschottdorf's question (I think he's right)

**Comment by mrtracy** [comment]
Yes, you are correct that initialProposal was not necessary.  I have updated this with Tobi's suggestion, everything appears to still be in working order.

**Comment by tamird** [comment]
:shipit:

**Pull request merged**